include/up-cpp/datamodel/builder/Uuid.h

```diff
@@ -0,0 +1,154 @@
+	static UuidBuilder getBuilder();
+
+	/// @brief Get a UuidBuilder in the test mode.
+	///
+	/// The testing mode of UuidBuilder allows for the time and random sources
+	/// to be replaced such that deterministic tests can be written in
+	/// situations where the normal behavior of UuidBuilder would interfere.
+	///
+	/// @remarks The provided UuidBuilder starts with an identical state to
+	///          one returned by getBuilder() with one difference: it will
+	///          allow customization of its behavior through the `withX()`
+	///          interfaces.
+	///
+	/// @returns A UuidBuilder in the test mode.
+	static UuidBuilder getTestBuilder();
+
+	/// @brief Sets the time source for a UuidBuilder in test mode.
+	///
+	/// @post All built UUIDs will use the provided function to get time
+	///       values instead of calling `std::chrono::system_clock::now()`
+	///
+	/// @note This can only be used with a UuidBuilder created with the
+	///       getTestBuilder() interface.
+	///
+	/// @param A callable that returns a time_point. Will be called whenever
+	///        build() is called to populate the time field in the built UUID.
+	///
+	/// @throws std::domain_error If called on a non-test UuidBuilder.
+	/// @returns A reference to thus UuidBuilder.
+	UuidBuilder& withTimeSource(
+	    std::function<std::chrono::system_clock::time_point()>&&);
+
+	/// @brief Sets the random value source for a UuidBuilder in test mode.
+	///
+	/// @post All built UUIDs will use the provided function to get random
+	///       values instead of using a true random source.
+	///
+	/// @note This can only be used with a UuidBuilder created with the
+	///       getTestBuilder() interface.
+	///
+	/// @param A callable that returns a uint64_t. Will be called whenever
+	///        build() is called in place of the default random value source.
+	///
+	/// @throws std::domain_error If called on a non-test UuidBuilder.
+	/// @returns A reference to thus UuidBuilder.
+	UuidBuilder& withRandomSource(std::function<uint64_t()>&&);
+
+	/// @brief Replaces the UuidSharedState with a new, non-shared, default
+	///        state.
+	///
+	/// @post The globally shared UUID state will no longer be used for
+	///       building UUIDs from this instance.
+	/// @post This instance will have the default values in its (non)shared
+	///       state.
+	///
+	/// @note This can only be used with a UuidBuilder created with the
+	///       getTestBuilder() interface.
+	///
+	/// @throws std::domain_error If called on a non-test UuidBuilder.
+	/// @returns A reference to thus UuidBuilder.
+	UuidBuilder& withIndependentState();
+
+	/// @brief Creates a uProtocol UUID based on the builder's current state.
+	///
+	/// @remarks As part of the UUID v7/v8 spec, there is a shared state for
+	///          all UUID builders within a process. Test builders can override
+	///          this with the withIndependentState() interface.
+	v1::UUID build();
```

Comment:
So the implementation of this needs to be done in a thread-safe manner, since threads A and B could be calling this at "the same time", right?

We have a test in `up-rust` to ensure that this is the case and internally implemented this via [an atomic and Compare And Swap (CAS)](https://github.com/eclipse-uprotocol/up-rust/blob/04c9518bb451bcab233be3f762ec68e99ea67494/src/uuid/uuidbuilder.rs#L81).

Reply:
Yes, the implementation will need to be thread-safe. That can either be achieved through *careful* use of `std::atomic` or by guarding critical sections with a mutex. If it can be done safely with atomic CAS, then that would be the preferred implementation.